File: getting_started/insertion_sort_decreasing_order/insertion_sort_dec_python/insertion_sort_dec.py
# Implementation of the insertion Sort algorithm in an decreasing order
# Author : Abdoulaye Diallo
# Complexity O(n²)

def insertionSortDecreasing(a):
    for j in range(1, len(a)):
        key = a[j]
        i = j-1
        while i > -1 and a[i] < key:
            a[i+1] = a[i]
            i = i-1
        a[i+1] = key
    return  a


a = [5, 2, 4, 6, 1, 3]

print("{}".format(insertionSortDecreasing(a)))

File: getting_started/insertion_sort_decreasing_order/insertion_sort_dec_cpp/insertion_sort_dec.cpp
/**
 * Implementation of the insertion sort algorithm in an decreasing order
 * Author : Abdoulaye Diallo
 * Complexity O(n²)
 */


#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void insertionSortDecreasing(vector<int> & Arr);

int main() {
    vector<int> Arr = {31, 41, 59, 26, 41, 58};

    insertionSortDecreasing(Arr);

    for (int i = 0; i < Arr.size(); i++) {
	cout << Arr[i] << " ";
    }

    cout << endl;
    
    return 0;
}

void insertionSortDecreasing(vector<int> & Arr) {
    int n = (int)Arr.size();
    int i, key;
    for (int j = 1; j < n; ++j) {
	key = Arr[j];
	i = j-1;

	while (i > -1 && Arr[i] < key) {
	    Arr[i+1] = Arr[i] ;
	    i = i-1;
	}
	Arr[i+1] = key;
    }
}
